Add Comet ML integration

caleb-kaiser opened the pull request:
## Motivation
I work on many of Comet ML's open source efforts, and recently, we've had a number of users request an official integration with Optuna. I implemented a straightforward integration between Comet and Optuna that provides a similar API to other experiment tracking platforms who've integrated with Optuna previously. Happy to incorporate any feedback the team might 
have! My assumption is that I should open a separate PR for documentation only after/if the team decides to approve the integration PR, but please let me know if that is not the case.

## Description of the changes
- Adds a `CometCallback` class in `optuna_integration/comet.py` that can be passed in as a callback to `study.optimize()`.
- Exposes a `@CometCallback.track_in_comet` decorator that allows users to decorate objective functions and directly access their Trial's Comet Experiment.
- Supports single and multi-objective optimization, with default support for Comet's Optimizer Reports for the single object use-case.
- Adds comprehensive unit tests in `tests/test_comet.py`, in keeping with this repository's testing guidelines

## Testing

To see the integration in action, you can run the following code (note you'll need to have Comet installed and setup with a free API key, for which there are instructions here: https://www.comet.com/docs/v2/guides/getting-started/quickstart/ )

```
import optuna
from optuna_integration.comet import CometCallback

from sklearn.datasets import load_iris
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, top_k_accuracy_score
from sklearn.model_selection import train_test_split
import random

study = optuna.create_study(study_name="optuna_distributed_single_example", directions=['maximize'])
comet =  CometCallback(study, project_name="optuna-distributed-single-example", metric_names=['score'])

@comet.track_in_comet()
def objective(trial):
    data = load_iris()
    x_train, x_valid, y_train, y_valid = train_test_split(data["data"], data["target"])

    params = {
        "min_samples_leaf": trial.suggest_int("min_samples_leaf", 2, 10),
        "max_depth": trial.suggest_int("max_depth", 5, 20),
        "min_samples_split": trial.suggest_int("min_samples_split", 2, 10),
    }

    clf = RandomForestClassifier(**params)
    clf.fit(x_train, y_train)
    pred = clf.predict(x_valid)
    score = accuracy_score(y_valid, pred)
    trial.experiment.log_other("foo", "bar")

    return score

db_path = "sqlite:///example.db"
study.optimize(objective, n_trials=30, callbacks=[comet])
```

Attached are screenshots of the output of the above code, which you can also view here: https://www.comet.com/ckaiser/optuna-distributed-single-example

Optimizer report view:
![image](https://github.com/optuna/optuna-integration/assets/42076840/2c6368be-1c79-401c-977b-c7110d16e86d)

Experiment dashboard view:
![image](https://github.com/optuna/optuna-integration/assets/42076840/9e368260-672e-481e-83f8-266abfecb473)

Single experiment view:
![image](https://github.com/optuna/optuna-integration/assets/42076840/94a0b5dd-2aab-4a52-8ab2-ee679e841c35)

nzw0301 reviewed (commented):
Awesome! Could you add `comet_ml` to https://github.com/optuna/optuna-integration/blob/main/pyproject.toml#L54 to fix the CI error? Note that the listing order is alphabetical.

caleb-kaiser commented:
> comet_ml

@nzw0301 no problem! I added Comet to the `pyproject.toml` and fixed an issue with one of the tests. There's a linting problem with `chainernm.py`, but since I didn't modify that file in my PR, I didn't want to overstep and change anything there. It seems like all other tests are passing though :)

nabenabe0928 commented:
@caleb-kaiser Thank you for the PR! We are checking if there are any issues regarding the license of comet-ml, so it may still take some time till we start the PR review!

caleb-kaiser commented:
@nabenabe0928 Sorry for any confusion! I work for Comet ML, and I'm part of the engineering team that maintains comet-ml. It is no problem for us to place the integration in Optuna. It does not cause any issues relating to licensing. Is there anything specific about comet-ml or licensing I can answer for you  that might help?

contramundum53 commented:
This PR is not stale.

caleb-kaiser commented:
@nabenabe0928 Thank you for the feedback/your patience on this. We've just completed the process of re-licensing the Comet SDK to be MIT licensed! Does this address any concerns about integrating Optuna and Comet?

nabenabe0928 commented on pyproject.toml:
```diff
@@ -55,6 +55,7 @@ all = [
   "catalyst",
   "catboost>=0.26; sys_platform!='darwin'",
   "catboost>=0.26,<1.2; sys_platform=='darwin'",
+  "comet_ml",
```
```suggestion
  "comet_ml>=3.39.3",
```

nabenabe0928 reviewed (changes requested):
Thank you for the license update!
As the MIT license is effective only after v3.39.3, could you add the version constraint?

I added the version constraint to the toml file as a suggestion!

HideakiImamura commented:
Thanks for the update and adding the license! Could you resolve the conflict and fix CIs?

nzw0301 commented:
Ping @HideakiImamura @nabenabe0928

HideakiImamura commented:
@caleb-kaiser Could you fix the CI?

Lothiraldan commented:
Hi @HideakiImamura, I'm a colleague from @caleb-kaiser. We got the fix ready internally but @caleb-kaiser didn't had the time to sent it before his vacation, sorry about that. He will send it early next week.

caleb-kaiser commented:
Hi @HideakiImamura ! The changes have been made, and it looks like everything is passing!

HideakiImamura reviewed (approved):
Sorry for the late. LGTM!

HideakiImamura commented:
I'm so sorry that there is a conflict in `.github/workflows/checks.yml`. Could you resolve it? I will merge this PR after the conflict is resolved immediately.

caleb-kaiser commented:
@HideakiImamura it looks like all tests are passing now :)

nzw0301 commented on optuna_integration/comet/comet.py:
```diff
@@ -0,0 +1,207 @@
+        trial_experiment.log_other("trial_number", trial.number)
+        trial_experiment.add_tag(f"trial_number_{trial.number}")
+
+        print(type(trial))
```
```suggestion
```

nzw0301 commented on optuna_integration/comet/comet.py:
```diff
@@ -0,0 +1,207 @@
+            API = comet_ml.api.API()
+            self._workspace = API.get_default_workspace()
+
+        # APIExperiment associated with the Optuna Study
```
Optuna coding style prefers using `.` at the end of sentence.
https://github.com/optuna/optuna/wiki/Coding-Style-Conventions#block-and-inline-comments-are-supposed-to-be-sentences
```suggestion
        # APIExperiment associated with the Optuna Study.
```

nzw0301 commented on optuna_integration/comet/comet.py:
```diff
@@ -0,0 +1,207 @@
+import json
+from typing import Any
+from typing import Callable
+from typing import Dict
```
Oning issue https://github.com/optuna/optuna/issues/4508 prefers `dict` instead of `typing.dict`. Similarly
`Union` -> `|`, and `Optional[SOME_TYPE]`-> `SOME_TYPE | None`.

https://github.com/optuna/optuna/issues/4508#issuecomment-1465563243

nzw0301 commented on optuna_integration/comet/comet.py:
````diff
@@ -0,0 +1,207 @@
+    - project_name (str) - Optional: The name of the project in Comet ML where the experiment will be logged. Defaults to "general"
+    - metric_names ([str]) - Optional: A list of the names of your objective metrics.
+
+    Usage:
+    ```
+    study = optuna.create_study(directions=["minimize", "maximize"])
+    comet_callback = CometCallback(study, metric_names=["accuracy", "top_k_accuracy"],
+                                     project_name="your_project_name", workspace="your_workspace")
+    study.optimize(your_objective_function, n_trials=100, callbacks=[comet_callback])
+    ```
+
+    Note:
+    The callback checks for an existing Comet Experiment key in the study's user attributes. If present, it initializes
+    an ExistingExperiment; otherwise, it creates a new APIExperiment and stores its key in the study for future reference.
+
+    You will need a Comet API key to log data to Comet.
+
+    You can also log extra data directly to your Trial's Experiment via the objective function by using
+    the @CometCallback.track_in_comet decorator, which exposes an `experiment` property on your trial, like so:
+    ```
+    study = optuna.create_study(directions=["minimize", "maximize"])
+    comet_callback = CometCallback(study, metric_names=["accuracy", "top_k_accuracy"],
+                                     project_name="your_project_name", workspace="your_workspace")
+
+    @comet_callback.track_in_comet()
+    def your_objective(trial):
+        trial.experiment.log_other("foo", "bar")
+        # Rest of your objective function...
+
+    study.optimize(your_objective, n_trials=100, callbacks=[comet_callback])
````
Could you follow the same format of other integration such as https://github.com/optuna/optuna-integration/blob/main/optuna_integration/wandb/wandb.py for documentation?

nzw0301 reviewed (commented):
Let me leave comments related coding styles and ongoing issue.

HideakiImamura reviewed (approved):
LGTM.

Let me merge this PR. The comments from @nzw0301 will be addressed in the follow-up PR.
